File: app/src/main/java/tv/spacedentist/android/SDModule.java
package tv.spacedentist.android;

import com.google.android.gms.cast.framework.CastContext;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import tv.spacedentist.android.chromecast.SDChromecastManager;
import tv.spacedentist.android.util.SDLogger;
import tv.spacedentist.android.util.SDLoggerAndroid;
import tv.spacedentist.android.util.SDLoggerNull;

@Module
public class SDModule {

    private final SDApplication mApplication;

    public SDModule(SDApplication application) {
        mApplication = application;
    }

    @Provides
    @Singleton
    SDLogger provideLogger() {
        return BuildConfig.DEBUG ? new SDLoggerAndroid() : new SDLoggerNull();
    }

    @Provides
    CastContext provideCastContext() {
        return CastContext.getSharedInstance(mApplication);
    }

    @Provides
    @Singleton
    SDChromecastManager provideChromecastManager() {
        return new SDChromecastManager(mApplication.getComponent());
    }

    @Provides
    @Singleton
    SDNotificationManager provideNotificationManager() {
        return new SDNotificationManager(mApplication, mApplication.getComponent());
    }
}

File: app/src/main/java/tv/spacedentist/android/SDModule.kt
package tv.spacedentist.android

import com.google.android.gms.cast.framework.CastContext
import dagger.Module
import dagger.Provides
import tv.spacedentist.android.chromecast.SDChromecastManager
import tv.spacedentist.android.util.SDLogger
import tv.spacedentist.android.util.SDLoggerAndroid
import javax.inject.Singleton

@Module
open class SDModule(private val mApplication: SDApplication?) {

    @Provides
    @Singleton
    internal open fun provideLogger(): SDLogger {
        return if (BuildConfig.DEBUG) SDLoggerAndroid() else object: SDLogger {}
    }

    @Provides
    internal open fun provideCastContext(): CastContext {
        return CastContext.getSharedInstance(mApplication!!)
    }

    @Provides
    @Singleton
    internal fun provideChromecastManager(): SDChromecastManager {
        return SDChromecastManager(mApplication!!.component)
    }

    @Provides
    @Singleton
    internal open fun provideNotificationManager(): SDNotificationManager {
        return SDNotificationManager(mApplication!!, mApplication.component)
    }
}
